pylint/checkers/similar.py

```diff
@@ -134,35 +434,73 @@ def _display_sims(self, sims):
             )
         )
 
-    def _find_common(self, lineset1, lineset2):
-        """find similarities in the two given linesets"""
-        lines1 = lineset1.enumerate_stripped
-        lines2 = lineset2.enumerate_stripped
-        find = lineset2.find
-        index1 = 0
-        min_lines = self.min_lines
-        while index1 < len(lineset1):
-            skip = 1
-            num = 0
-            for index2 in find(lineset1[index1]):
-                num_lines_with_content = 0
-                for num, ((_, line1), (_, line2)) in enumerate(
-                    zip(lines1(index1), lines2(index2))
-                ):
-                    if line1 != line2:
-                        if num_lines_with_content > min_lines:
-                            yield num, lineset1, index1, lineset2, index2
-                        skip = max(skip, num)
-                        break
-                    if re.match(REGEX_FOR_LINES_WITH_CONTENT, line1):
-                        num_lines_with_content += 1
-                else:
-                    # we may have reached the end
-                    num += 1
-                    if num_lines_with_content > min_lines:
-                        yield num, lineset1, index1, lineset2, index2
-                    skip = max(skip, num)
-            index1 += skip
+    def _find_common(
+        self, lineset1: "LineSet", lineset2: "LineSet"
+    ) -> Generator[
+        Tuple[
+            int, "LineSet", LineNumber, LineNumber, "LineSet", LineNumber, LineNumber
+        ],
+        None,
+        None,
+    ]:
+        """
+        Find similarities in the two given linesets.
+
+        This the core of the algorithm.
+        The idea is to compute the hashes of a minimal number of successive lines of each lineset and then compare the hashes.
+        Every match of such comparison is stored in a dict that links the couple of starting indices in both linesets to
+        the couple of corresponding starting and ending lines in both files.
+        Last regroups all successive couples in a bigger one. It allows to take into account common chunk of lines that have more
+        than the minimal number of successive lines required.
+        """
+        hash_to_index_1: HashToIndex_T
+        hash_to_index_2: HashToIndex_T
+        index_to_lines_1: IndexToLines_T
+        index_to_lines_2: IndexToLines_T
+        hash_to_index_1, index_to_lines_1 = hash_lineset(lineset1, self.min_lines)
+        hash_to_index_2, index_to_lines_2 = hash_lineset(lineset2, self.min_lines)
+
+        hash_1: FrozenSet[LinesChunk] = frozenset(hash_to_index_1.keys())
+        hash_2: FrozenSet[LinesChunk] = frozenset(hash_to_index_2.keys())
+
+        common_hashes: Iterable[LinesChunk] = sorted(
+            list(hash_1 & hash_2), key=lambda m: hash_to_index_1[m][0]
```

Comment:
```suggestion
            hash_1 & hash_2, key=lambda m: hash_to_index_1[m][0]
```
You can use a generator here